Spec/Auth.swift

```diff
@@ -4030,24 +3982,45 @@ override class var defaultTestSuite : XCTestSuite {
                     }
                 }
             }
-        }
+enum TestCase_ReusableTestsTestTokenRequestFromJson {
+case accepts_a_string__which_should_be_interpreted_as_JSON
+case accepts_a_NSDictionary
+}
+
+                func reusableTestsTestTokenRequestFromJson(_ json: String, testCase: TestCase_ReusableTestsTestTokenRequestFromJson, context: (beforeEach: (() -> ())?, afterEach: (() -> ())?), check: @escaping (_ request: ARTTokenRequest) -> Void) {
+                    func test__accepts_a_string__which_should_be_interpreted_as_JSON() {
+context.beforeEach?()
 
-        describe("TokenRequest") {
-            // TE6
-            describe("fromJson") {
-                func reusableTestsTestTokenRequestFromJson(_ json: String, check: @escaping (_ request: ARTTokenRequest) -> Void) {
```

Comment:
I'll explain how we handle `reusableTests…` functions, which are functions that we specially marked in #1225 as functions that contain calls to `context` / `it`. The migrator recognises these functions and does the following:

### How we transform the function declaration

In this example, that'd be the `func reusableTestsTestTokenRequestFromJson` declaration.

1. It flattens all of the `it` calls contained within the function declaration into a list of individual function declarations (e.g. `test__accepts_a_NSDictionary` here).
2. It creates a `TestCase_…` enum which describes each of the test cases contained within the function declaration (see `TestCase_ReusableTestsTestTokenRequestFromJson` here).
3. It adds a `testCase` param to the function declaration, allowing the caller to specify which test case they want to execute, and it adds a `switch` statement to the body of the function declaration which invokes the specified test case

### How we translate the function call

Reply:
It's also worth pointing out that the `reusableTests…` functions are the only place where the behaviour of the migrated tests differs from that of the original Quick tests. In the original Quick specs, the local variables inside the `reusableTests…` function declaration are only initialized once. In the migrated specs, they're initialized once for _each test case within the function declaration_. If I had unlimited time, I would have fixed this, but I didn't consider it worthwhile since it doesn't seem to have introduced any issues.

Reply:
Sometimes repeating code is not a crime. Especially testing code. I would just unwrap it into individual `test_*` functions. Just a suggestion.

Reply:
> you'd expand the entire reusableTests function, removing the function declaration entirely

Yes.

> which is not bad just because it's repetition, but also because it'll make this diff a lot harder to review.

Agreed. It's already there.